library/fsio.c

```diff
@@ -243,172 +240,144 @@ int mbedtls_fseek( mbedtls_file_t *stream, long offset, int whence )
-    return( dir );
+    CHECK_STATUS( mbedtls_serialize_push_buffer( path, strlen( path ) + 1 ), 0, MBEDTLS_DIR_INVALID );
+    CHECK_STATUS( mbedtls_serialize_execute( MBEDTLS_SERIALIZE_FUNCTION_DOPEN ), 0, MBEDTLS_DIR_INVALID );
+    CHECK_STATUS( mbedtls_serialize_pop_int32( ( uint32_t * )&dir_id ), 0, MBEDTLS_DIR_INVALID );
+    return( dir_id );
 }
 
 /**
  * \brief           Read dir entry (file, dir etc.).
  *
- * \param dir       Pointer to directory handle.
+ * \param dir       Dir handle of type mbedtls_dir_t.
  * \param dirent    Out buffer for directory entry name.
  *                  Upto 255 character long name can be returned.
  * \param size      Out buffer length.
  * \param type      Entry type.
  *
  * \return          0 for success. Non zero for failure.
  */
-int mbedtls_readdir( mbedtls_dir_t * dir, char * dirent, uint32_t size,  uint32_t * type )
+int mbedtls_readdir( mbedtls_dir_t dir, char * dirent, uint32_t size,  uint32_t * type )
 {
-    int status;
     size_t len = 0;
 
-    mbedtls_serialize_push_int32( PTR_TO_INT( dir ) );
-    mbedtls_serialize_push_int32( size ); /* Send entry size for validation */
-    status = mbedtls_serialize_execute( MBEDTLS_SERIALIZE_FUNCTION_DREAD );
-    if ( status == 0 )
-    {
-        mbedtls_serialize_pop_int32( (uint32_t *)type );
-        mbedtls_serialize_pop_buffer( dirent, size, &len );
+    CHECK_STATUS( mbedtls_serialize_push_int32( dir ), 0, -1 );
+    CHECK_STATUS( mbedtls_serialize_push_int32( size ), 0, -1 );  /* Send entry size for validation */
+    CHECK_STATUS( mbedtls_serialize_execute( MBEDTLS_SERIALIZE_FUNCTION_DREAD ), 0, -1 );
+    CHECK_STATUS( mbedtls_serialize_pop_int32( (uint32_t *)type ), 0, -1 );
+    CHECK_STATUS( mbedtls_serialize_pop_buffer( dirent, size, &len ), 0, -1 );
 
-        /* Map serialize dir entry type to fsio abstraction type */
-        switch ( *type )
-        {
-            case MBEDTLS_SERIALIZE_DT_BLK:
-                *type = MBEDTLS_FSIO_DT_BLK;
-                break;
-            case MBEDTLS_SERIALIZE_DT_CHR:
-                *type = MBEDTLS_FSIO_DT_CHR;
-                break;
-            case MBEDTLS_SERIALIZE_DT_DIR:
-                *type = MBEDTLS_FSIO_DT_DIR;
-                break;
-            case MBEDTLS_SERIALIZE_DT_FIFO:
-                *type = MBEDTLS_FSIO_DT_FIFO;
-                break;
-            case MBEDTLS_SERIALIZE_DT_LNK:
-                *type = MBEDTLS_FSIO_DT_LNK;
-                break;
-            case MBEDTLS_SERIALIZE_DT_REG:
-                *type = MBEDTLS_FSIO_DT_REG;
-                break;
-            case MBEDTLS_SERIALIZE_DT_SOCK:
-                *type = MBEDTLS_FSIO_DT_SOCK;
-                break;
-            case MBEDTLS_SERIALIZE_DT_UNKNOWN:
-            default:
-                *type = MBEDTLS_FSIO_DT_UNKNOWN;
-                break;
-        }
+    /* Map serialize dir entry type to fsio abstraction type */
+    switch ( *type )
+    {
+        case MBEDTLS_SERIALIZE_DT_BLK:
```

Comment:
When using abstraction we don't want to directly include ```dirent.h```. Hence mapping those constants to equivalent types defined by the abstraction.

Reply:
@mazimkhan I know why we aren't using constants from `dirent.h`. What I don't understand is why we're defining two sets of constants, `MBEDTLS_FSIO_xxx` from `fsio.h` and `MBEDTLS_SERIALIZE_DT_xxx` from `serialize.h`, which cover the exact same set of possibilities.

Reply:
Definitions in ```fsio.h``` are guarded under ```MBEDTLS_FS_IO```. Building library is fine as that uses ```config.h```. However, ```serialize.h``` is also included in ```frontend.c``` that is built with rather less config. Hence wanted to keep ```serialize.h``` as independent as it could. But yes it is possible to include ```fsio.h``` in it.

Reply:
The frontend calls platform functions anyway to execute the functions, so it has to use `MBEDTLS_FS_IO`, and not only include the headers but also compile and link against the functions.

Reply:
It is possible. Before that,
"The frontend calls platform functions" 
I would not say it like this. frontend does not call platform functions. It implements them over the serialised interface when they are absent in the library. There is no need to turn On serialisation and frontend if they are available.

Now back to the original point. library and frontend are not compiled with same config. Consumers of ```serialise.h``` i.e. ```fsio.h``` and ```frontend.c``` are built under two different configs. Hence open to conflict in future. Hence I am not very happy mixing ```fsio.h``` inside ```serialise.h```. But yes it is possible.

Reply:
Also if there is any code in the library that is only consumed by frontend.c then it should not in the library. I know there could be some re-usability. But if it just about std C functions then I will not mix library into frontend.c. For the sake of simplicity in code and build system.